source/fab/tools/tool_repository.py

```diff
@@ -117,26 +120,50 @@ def set_default_compiler_suite(self, suite: str):
         '''
         for category in [Category.FORTRAN_COMPILER, Category.C_COMPILER,
                          Category.LINKER]:
-            all_members = [tool for tool in self[category]
-                           if tool.suite == suite]
-            if len(all_members) == 0:
+            # Now sort the tools in this category to have all tools with the
+            # right suite at the front. We use the stable sorted function with
+            # the key being tool.suite != suite --> all tools with the right
+            # suite use False as key, all other tools True. Since False < True
+            # this results in all suite tools to be at the front of the list
+            self[category] = sorted(self[category],
+                                    key=lambda x: x.suite != suite)
+            if len(self[category]) > 0 and self[category][0].suite != suite:
                 raise RuntimeError(f"Cannot find '{category}' "
                                    f"in the suite '{suite}'.")
-            tool = all_members[0]
-            if tool != self[category][0]:
-                self[category].remove(tool)
-                self[category].insert(0, tool)
 
-    def get_default(self, category: Category):
-        '''Returns the default tool for a given category, which is just
-        the first tool in the category.
+    def get_default(self, category: Category,
+                    mpi: Optional[bool] = None):
+        '''Returns the default tool for a given category. For most tools
+        that will be the first entry in the list of tools. The exception
+        are compilers and linker: in this case it must be specified if
+        MPI support is required or not. And the default return will be
+        the first tool that either supports MPI or not.
```

Comment:
Fair point, I have half a dozen alternatives, here the few that I think are actually sensible:
1. We add separate `get_compiler`/`get_linker` methods (which take mpi as parameter), nothing else does.
2. We abort if a value for mpi is specified for any non-compiler/linker too. Or we turn this into kargs, meaning we can add tool-specific flags in the future, and for now only accept MPI for compiler/linker.
3. We add the `BuildConfig` as argument instead of mpi only. We can access the mpi status this way, and whatever any tool might need in the future could be done (admittedly I can't think of anything useful here :)  ... maybe, and I am clutching at straws here, a tool could access the number of cores to use in a build, if a tool itself can run in parallel??)

Would one of these solution work for you? Or do you have a better idea?

Reply:
Of the options given I think I prefer number 1.

I don't like exceptions on invalid arguments, I would rather the arguments were valid. I'm also wary of using `kargs` as it throws away all the advantages of static call signature checking. Also of static type analysis.

Throwing `BuildConfig` at the problem also suffers from removing static checking.

One possible solution was hinted at earlier. The `Category` enumeration could take an MPI argument for compiler and linker tools. I think this would work and would be limited to just the categories which care about the concept of MPI.

When calling: `tool_repo.get_default(FortranCompiler(mpi=True))`. In the method: `if isinstance(category, FortranCompiler): needs_mpi = category.needs_mpi()`.

Reply:
I don't understand what you mean with Category .. enums are afaik not configurable. We could add a method `get_fortran(mpi)` etc to Category that would return FORTRAN_MPI or FORTRAN_NO_MPI, but given that we need this three times (Fortran, C, linker), this feels like a huge amount of unnecessary code (given the current size of category.py). Plus, if we test if a tool is really a Fortran compiler, we then need to test for two potential values :(

Adding get_compiler/get_linker also feels odd when I tried to do this: the `get_compiler`  would still need to have a category (to get Fortran or c), or we add get_c_compiler and get_fortran _compiler. This then also need to be duplicated, since we need the same functions then in ToolBox and ToolRepository.

This all gets very messy if we want to add different language (C++ anyone?) - atm that might be reasonable easy to do (add the allowed suffixes, add a category, and pick C or C++ compiler based on suffix,... done .. mostly, I am sure I missed something :) But compared with then having to add get_cpp_compiler twice, ...) - feels wrong. I think that would couple the language way too much into otherwise generic tool handling classes.

My suggestion would then be to add the config object as parameter - either optional (I can't think of many other tools needing this :) ), or mandatory.

Reply:
I had a deeper look into `Enum` and you are right. They can take parameters but only across all members of the `Enum`. i.e. every option carries the same parameters.

I've raised #350 to consider an alternative approach.